Make DeepGPs copyable (ish)

uri-granta opened the pull request:
Add support for copying DeepGPs by making DeepGaussianProcess accept a closure so we don't have to copy the architecture (something gpflux doesn't support).

Also fix issue with deep ensembles failing to copy when the fit args doesn't include a callbacks key.

Commit:
Track state in integration state

Commit:
Handle callback-less optimizers

hstojic commented:
not great that we have to use a closure, but its good to have a solution, it seems good otherwise, I don't have cosntructive comments except that it would be good to mention it in the notebook as well

Commit:
Typo

Commit:
Update deepgp notebook

Commit:
Update track state error message

uri-granta commented:
I've added an explanation to the deepgp notebook and changed it so it no longer uses `track_state=False`. I've also updated the generic exception message which gets printed when track_state fails.

Commit:
Format

hstojic reviewed (approved):
looks good to me!
